### ISSUE TITLE
Missing code actions

### ISSUE THREAD

**Issue by jacg** [emacs-lsp/lsp-ui#319]
I'm not sure whether this is a problem in lsp-ui, or the server that I am using, but I have a concrete starting point in lsp-ui ...

My server is [ghcide](https://github.com/digital-asset/ghcide). Here is a source file that can serve for illustrative testing purposes:

````
{-# OPTIONS_GHC -Wmissing-signatures #-}

module Testing where

foo = True
````

3 observations:

1. **FAIURE**: `foo` on line 5 is marked as a warning whith message `Top-level binding with no type signature: foo :: Bool`. In such situations `ghcide` proposes an action for adding the type signature. Even though code action display has been switched on, no code action is shown in the sideline.

2. **FAILURE**: Running `lsp-ui-sideline-apply-code-actions` on line 5 results in `user-error: No code actions on the current line`

3. **SUCCESS**: Running `lsp-execute-code-action` results in the type signature being inserted.

However, if I modify `lsp-ui-sideline--run`

````diff
-(when (and lsp-ui-sideline-show-code-actions (or (lsp--capability "codeActionProvider")
-                                                 (lsp--registered-capability "textDocument/codeAction")))
+(when (and lsp-ui-sideline-show-code-actions))
````
in other words, showing code actions in the sideline if they are requested, without checking related server capabilities, then the above observations change to

1. **SUCCESS**: The add-signature code action is shown in the sideline on line 5.
2. **SUCCESS**: `lsp-ui-sideline-apply-code-actions` first proposes, then correctly applies the add-signature action.
3. **SUCCESS**: `lsp-execute-code-action` continues to work correctly as before.

I'm not intimately familiar with LSP, but it looks like there's some miscommunication between `lsp-ui` and `ghcide` about whether `ghcide` provides code actions.

Can you suggest how to track down whether this is a problem with `lsp-ui`, `ghcide` or something else? and what I can do to help fix it?

**Comment by yyoncho** [comment]
What is the value of lsp-ui-sideline-update-mode ?

**Comment by jacg** [comment]
> What is the value of lsp-ui-sideline-update-mode ?

`line`

**Comment by yyoncho** [comment]
Sync to latest lsp-ui and the issue will disappear see: https://github.com/emacs-lsp/lsp-ui/commit/fba17d4127fa768ca022f405a7a363621b73fc00

**Comment by jacg** [comment]
Running a fresh clone of `lsp-ui` (current `master` c8fa40c0f9c65877d1cabe1739e5f787adb24898, which includes fba17d4) the issue has **not** disappeared.

**Comment by yyoncho** [comment]
Hm, can you check what is the kind of the code action in the client/server logs? 

By default, the sideline is displaying only the code actions that match `lsp-ui-sideline-actions-kind-regex`

**Comment by jacg** [comment]
The kind is `quickfix` which features in `lsp-ui-sideline-action-kind-regex`.

Here it is in its full glory:

````
[Trace - 12:37:14 pm] Received response 'textDocument/codeAction - (134)' in 1ms.
Result: [
  {
    "title": "add signature: foo :: Bool",
    "diagnostics": [
      {
        "message": "/home/jacek/test-lsp-ui/Testing.hs:12:1: warning:\n    Top-level binding with no type signature: foo :: Bool",
        "source": "typecheck",
        "range": {
          "end": {
            "character": 3,
            "line": 11
          },
          "start": {
            "character": 0,
            "line": 11
          }
        },
        "severity": 2
      }
    ],
    "kind": "quickfix",
    "edit": {
      "changes": {
        "file:///home/jacek/test-lsp-ui/Testing.hs": [
          {
            "newText": "foo :: Bool\n",
            "range": {
              "end": {
                "character": 0,
                "line": 11
              },
              "start": {
                "character": 0,
                "line": 11
              }
            }
          }
        ]
      }
    }
  }
]
````

**Comment by yyoncho** [comment]
Can you include full client/server communication up to that point? It seems like your issue might not be related to the recent lsp-ui-sideline changes.

**Comment by jacg** [comment]
Hang on ... there are no code action messages coming in, unless I run `lsp-execute-code-action`.

**Comment by jacg** [comment]
> Can you include full client/server communication up to that point?

OK, I'll try to get a clean session, so that it's not too big ...

**Comment by yyoncho** [comment]
@jacg as you initial comment states, it seems like the server does not report that it supports code actions and lsp-ui-sideline ignores them. I checked the code in lsp-execute-code-action and it does not do such check which explains why it behaves in a different way.

**Comment by jacg** [comment]
https://pastebin.com/TTSZZTtM

The variable started off being called `foo`. I took care not to request any code actions to be performed before changing its name to `xfoo`, at which point I immediately executed `lsp-execute-code-action`.

**Comment by jacg** [comment]
> it seems like the server does not report that it supports code actions

Can you help me come up with some non-interactive test that can verify this?

**Comment by yyoncho** [comment]
The issue is that the server does not declare that it supports code actions in ServerCapabilities - https://microsoft.github.io/language-server-protocol/specification so it should be reported in the language server repo.

**Comment by jacg** [comment]
> it should be reported in the language server repo.

That's clear, but can you give me any hints on how to verify this behaviour automatically? Ideally I'd want to write some minimal stand-alone client that connects to the server, and elicits some response from the server which can then be compared to what is expected.

At the moment I have to fire up emacs (ensuring `lsp` and friends are loaded, but irrelevant customizations are not), open some file, perform some actions in Emacs, etc. and then observe what happens on my screen and waffle on about it. It would be so much better if I could submit a fully-automated test.

As I'm not intimate with LSP, I was hoping you might give me some hints on how one might set up minimal, automated, verifiable communication with the server.

**Comment by yyoncho** [comment]
Submitting the response of initialize response will be sufficied for the server side team to understand and fix the issue.

If you want to reproduce it in emacs do you could something like that:

``` elisp
(add-hook 'lsp-after-initialize-hook (lambda ()
                                       (unless (or (lsp--capability "codeActionProvider")
                                                   (lsp--registered-capability "textDocument/codeAction"))

                                         (warn "codeActionProvider not present"))))
```

If you want to automate it outside of emacs - this is more work because you will have to create jsonrpc client. 

Closing the issue, since there is nothing we can do on our side. We could make lsp-mode consistent with lsp-ui and fail the call but that won't help your case.
